src/toast/observation_dist.py

```diff
@@ -164,170 +193,285 @@ def __init__(
+        if recv_slc is not None:
+            n_elem = 1
+            rs = list()
+            for slc in recv_slc:
+                rlen = slc.stop - slc.start
+                n_elem *= rlen
+                rs.append(rlen)
+            recv_shape.append(tuple(rs))
+        else:
+            recv_shape.append(None)
+        recv_counts[iproc] = n_elem
+        off += n_elem
+    recv_n_elem = off
+
+    # Set up the send / receive buffers
+    send_buf = buffer_class(send_n_elem)
+    recv_buf = buffer_class(recv_n_elem)
+
+    # Copy data into the send buffer
+    for iproc, send_slc in enumerate(send_info):
+        off = send_displ[iproc]
+        n_elem = send_counts[iproc]
+        if n_elem > 0:
+            send_buf[off : off + n_elem] = input[send_slc].flatten()
+
+    # Communicate data
+    comm.Alltoallv(
+        [
+            send_buf,
+            send_counts,
+            send_displ,
+            mpi_type,
+        ],
+        [
+            recv_buf,
+            recv_counts,
+            recv_displ,
+            mpi_type,
+        ],
     )
 
-    det_order = {y: x for x, y in enumerate(self.all_detectors)}
+    # Copy data to output
+    for iproc, recv_slc in enumerate(recv_info):
+        off = recv_displ[iproc]
+        n_elem = recv_counts[iproc]
+        if n_elem > 0:
+            output[recv_slc] = (
+                recv_buf[off : off + n_elem].array().reshape(recv_shape[iproc])
+            )
+
+    # Delete buffers
+    send_buf.clear()
+    del send_buf
+    recv_buf.clear()
+    del recv_buf
+    send_counts.clear()
+    del send_counts
+    send_displ.clear()
+    del send_displ
+    recv_counts.clear()
+    del recv_counts
+    recv_displ.clear()
+    del recv_displ
+
+
+def redistribute_data(old_dist, new_dist, shr_mgr, det_mgr, ivl_mgr, times=None):
+    """Helper function to redistribute data in an observation.
+
+    Given the old and new data distributions, redistribute all objects in the
+    shared, detdata, and intervals manager objects.  The input managers are cleared
+    upon return.  If times is None and there exist some intervals, a warning
+    will be given and the intervals will be deleted.
 
-    if newdist.comm_rank == 0:
-        # check that all processes have some data, otherwise print warning
-        for d in range(newdist.process_rows):
-            if len(newdist.dets[d]) == 0:
-                msg = "WARNING: process row rank {} has no detectors"
-                " assigned in new distribution.".format(d)
-                log.warning(msg)
-        for r in range(newdist.process_cols):
-            if newdist.samps[r][1] <= 0:
-                msg = "WARNING: process column rank {} has no data assigned "
-                "in new distribution.".format(r)
-                log.warning(msg)
+    Args:
+        old_dist (DistDetSamp):  The existing data distribution.
+        new_dist (DistDetSamp):  The new data distribution.
+        mgr (SharedDataMgr):  The manager instance.
```

Comment:
The doc string seems to be outdated and missing `shr_mgr, det_mgr, ivl_mgr`. `times` is misspelled.